### PR TITLE
Spark 457661 prepare locus meeting leave fetch

### PR DESCRIPTION
**Pull request by chburket**
## COMPLETES

https://jira-eng-gpk2.cisco.com/jira/browse/SPARK-457661

## This pull request addresses

We need a way to prepare the options needed to submit a locus meeting /leave. This will allow us to call the function from cantina and then simply call fetch with the prepared options when the browser/tab is closed.

## by making the following changes

Adds a new meeting.buildLeaveFetchRequestOptions() function that prepares all the options required to submit a /leave Locus request via fetch, but does not call the fetch.

Other supporting functions were also added, as well as some rearranging of source into reusable functions.

### Change Type

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [X] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation update
- [ ] Tooling change
- [ ] Internal code refactor

## The following scenarios where tested

Unit tests and manual testing with updated Cantina.

### I certified that

- [X] I have read and followed [contributing guidelines](https://github.com/webex/webex-js-sdk/blob/master/CONTRIBUTING.md#submitting-a-pull-request)
- [ ] I discussed changes with code owners prior to submitting this pull request

- [X] I have not skipped any automated checks
- [X] All existing and new tests passed
- [X] I have updated the documentation accordingly

### CODE REVIEW

File: packages/@webex/internal-plugin-metrics/src/call-diagnostic/call-diagnostic-metrics.util.ts [hunk 1]
```diff
@@ -218,16 +218,19 @@ export const prepareDiagnosticMetricItem = (webex: any, item: any) => {
  * @returns {any} the updated options object
  */
 export const setMetricTimings = (options) => {
-  if (options.body?.metrics) {
+  if (options.body && options.json) {
+    const body = JSON.parse(options.body);
+
     const now = new Date().toISOString();
-    options.body.metrics.forEach((metric) => {
+    body.metrics?.forEach((metric) => {
```
Comment (chburket) [inline packages/@webex/internal-plugin-metrics/src/call-diagnostic/call-diagnostic-metrics.util.ts:225]: This is a related fix I slipped in here.

File: packages/@webex/plugin-meetings/test/unit/spec/meeting/index.js [hunk 2]
```diff
@@ -5252,16 +5254,16 @@ describe('plugin-meetings', () => {
...
-              EVENT_TRIGGERS.MEETING_LOCUS_URL_UPDATE,
-              {'locusUrl': 'newLocusUrl/12345'}
+            TriggerProxy.trigger,
+            sinon.match.instanceOf(Meeting),
+            {
+              file: 'meeting/index',
+              function: 'setUpLocusSelfListener',
+            },
+            EVENT_TRIGGERS.MEETING_LOCUS_URL_UPDATE,
+            {locusUrl: 'newLocusUrl/12345'}
```
Comment (chburket) [inline packages/@webex/plugin-meetings/test/unit/spec/meeting/index.js:5264]: Unfortunately, there are a lot of prettier formatting changes.

File: packages/@webex/internal-plugin-metrics/test/unit/spec/new-metrics.ts [hunk 3]
```diff
@@ -115,7 +115,7 @@ describe('internal-plugin-metrics', () => {
       });
     });
 
-    describe('#setMetricTimingsAndFetch', () => {
+    describe.only('#setMetricTimingsAndFetch', () => {
```
Comment (arun3528) [inline packages/@webex/internal-plugin-metrics/test/unit/spec/new-metrics.ts:118]: need to remove this

File: packages/@webex/plugin-meetings/src/meeting/index.ts [hunk 4]
```diff
@@ -5908,6 +5908,34 @@ export default class Meeting extends StatelessWebexPlugin {
...
+   * Note: the $timings values will be wrong, but setRequestTimingsAndFetch() will
+   * properly adjust them before submitting.
+   *
+   * @public
+   * @param {Object} options leave options
+   * @param {String} options.resourceId the device with which to leave from, empty if just the computer
+   * @param {any} options.reason the reason for leaving
+   * @returns {Promise} resolves to options to be used with fetch
+   */
+  public buildLeaveFetchRequestOptions(options: {resourceId?: string; reason?: any} = {} as any) {
```
Comment (arun3528) [inline packages/@webex/plugin-meetings/src/meeting/index.ts:5932]: @chburket can you add _ to it for now , we don't have to expose this to the developer for now

**Review by arun3528 — COMMENTED** [review]
small comment

File: packages/@webex/plugin-meetings/src/meeting/index.ts (hunk 4 above)
Comment (chburket) [reply packages/@webex/plugin-meetings/src/meeting/index.ts:5932]: We *do* need to expose it.  I am going to use it immediately from Cantina.

**Review by torgeadelin — CHANGES_REQUESTED** [review]

File: packages/@webex/internal-plugin-metrics/src/call-diagnostic/call-diagnostic-metrics.util.ts [hunk 5]
```diff
@@ -218,16 +218,19 @@ export const prepareDiagnosticMetricItem = (webex: any, item: any) => {
...
       if (metric.eventPayload) {
         metric.eventPayload.originTime = {
           triggered: now,
```
[the remark this answers is not in the document]
Comment (torgeadelin) [reply packages/@webex/internal-plugin-metrics/src/call-diagnostic/call-diagnostic-metrics.util.ts:228]: okay, maybe add a comment why this is the case

**Review by torgeadelin — APPROVED** [review]

**Review by arun3528 — APPROVED** [review]

**Pull request merged**